src/metabase/lib/metadata/invocation_tracker.cljc

```diff
@@ -0,0 +1,80 @@
+(ns metabase.lib.metadata.invocation-tracker
+  (:require
+   [metabase.lib.metadata.protocols :as lib.metadata.protocols]))
+
+(def ^:private ^:dynamic *to-track-metadata-types*
+  "Set of metadata types to track.
+  Currently only `:metadata/card` is tracked for updating report_card.last_used_at.
+  See [[metabase.query-processor.middleware.update-used-cards/update-used-cards!]].
+
+  Making this dynamic for testing purposes."
+  #{:metadata/card})
+
+(defn- update-tracker-and-call-next-provider!
+  [tracker metadata-provider metadata-type id]
+  ;; we only have usage for metadata/card for now, so we only track it to save some overhead
+  (when (*to-track-metadata-types* metadata-type)
+    (swap! tracker update metadata-type (fn [item-ids]
+                                          (if (seq item-ids)
+                                            (conj item-ids id)
+                                            [id]))))
+  ((case metadata-type
```

Comment:
This feels like we can automate it nicely with a macro, but that can wait until the next time we need a thin "decorator" like this.